### ISSUE TITLE
Code in "password_providers" is loaded, even if disabled

### ISSUE THREAD

**Issue by sumpfralle** [matrix-org/synapse#10171]
### Description

I am (successfully) using a custom password provider for matrix-synapse:
```yaml
password_providers:
  - module: "grouprise.auth.matrix_synapse_auth_grouprise.GroupriseAuthProvider"
    config:
      enabled: true
      settings_filename: /etc/grouprise/settings.py
```

The above snippets works as expected.

But when I disable the password provider (`enabled: false`), the python module is still loaded and the above class (`GroupriseAuthProvider`) is still initialized.

In order to prevent matrix-synapse from initializing the class, I need to comment/remove the above configuration snippet completely.

### Steps to reproduce

- add a minimal (broken) configuration snippet to the synapse configuration:
```yaml
password_providers:
  - module: "foo.bar"
    config:
      enabled: false
```
- restart matrix-synapse
- take a look at the logs and discover the following lines:
```
Traceback (most recent call last):
  File "/usr/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/usr/lib/python3/dist-packages/synapse/app/homeserver.py", line 507, in <module>
    main()
  File "/usr/lib/python3/dist-packages/synapse/app/homeserver.py", line 502, in main
    hs = setup(sys.argv[1:])
  File "/usr/lib/python3/dist-packages/synapse/app/homeserver.py", line 333, in setup
    "Synapse Homeserver", config_options
  File "/usr/lib/python3/dist-packages/synapse/config/_base.py", line 727, in load_or_generate_config
    config_dict, config_dir_path=config_dir_path, data_dir_path=data_dir_path
  File "/usr/lib/python3/dist-packages/synapse/config/_base.py", line 749, in parse_config_dict
    data_dir_path=data_dir_path,
  File "/usr/lib/python3/dist-packages/synapse/config/_base.py", line 374, in invoke_all
    res[name] = getattr(config, func_name)(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/synapse/config/password_auth_providers.py", line 49, in read_conf
    ("password_providers", "<item %i>" % i),
  File "/usr/lib/python3/dist-packages/synapse/util/module_loader.py", line 49, in load_module
    module = importlib.import_module(module)
  File "/usr/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 965, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'foo'
```

Instead of the above behavior, I expected synapse to ignore the disabled *password_provider* completely (i.e. not importing the module).

### Version information

- **Version**: 1.28.0-1~bpo10+2
- **Install method**: Debian package (buster-backports)

**Comment by ShadowJonathan** [comment]
The code is not executed, but the module is still trying to be loaded, subtle difference.

**Comment by erikjohnston** [comment]
This is an interesting one: I can see an argument that we *should* still load the password provider module as it might be surprising if it breaks when setting `enabled: true`.

**Comment by richvdh** [comment]
the `config` block (including the `enabled` setting shown here) contains options for the module, so of course we have to load the module, so that we can pass the option in. *As it happens*, some modules (including [LdapAuthProvider](https://github.com/matrix-org/matrix-synapse-ldap3/blob/main/ldap_auth_provider.py)) require an `enabled` option to enable them, though personally I think that is a misfeature. There is certainly nothing that says all modules need to implement `enabled`, nor that it necessarily means the same to all modules.

In short, I think having Synapse inspect the supposedly module-specific options and apply magic behaviour if there is one called `enable` would be extremely confusing.

If you want a way to temporarily avoid loading the password provider, just rename `password_providers` to something like `_x_password_providers` that will be ignored by Synapse.

**Comment by erikjohnston** [comment]
(Oh, I missed that the enable flag was in the module config, rather than the Synapse config).

**Comment by richvdh** [comment]
I'm going to go ahead and close this then, since everyone seems agreed.

**Comment by gramakri** [comment]
That's interesting that 'enabled' is actually part of the module's config. It seems that starting synapse 1.37.0, the 'enabled' flag is ignored altogether for the LDAP module. Even if 'enabled' is false, it connects to the LDAP server (as seen from the logs). However, I don't find any changes that deprecates the 'enabled' flag in the LDAP module code itself (yet).

**Comment by gramakri** [comment]
If I have this:
```
password_providers:
  - module: "synapse.util.ldap_auth_provider.LdapAuthProvider"
    config:
      enabled: false
      uri: "ldap://ldap.example.com:389"
      start_tls: true
      base: "ou=users,dc=example,dc=com"
      attributes:
        uid: "username"
        mail: "mail"
        name: "username"
      bind_dn: "ou=users,dc=cloudron"
      bind_password: "password"
      filter: "(objectClass=posixAccount)"
```

then, I still see this in the logs:
```
2021-07-24 05:36:29,449 - synapse.rest.client.v1.login - 271 - INFO - POST-16 - Got login request with identifier: {'type': 'm.id.user', 'user': 'girish'}, medium: None, address: None, user: None
2021-07-24 05:36:29,449 - ldap_auth_provider - 107 - DEBUG - POST-16 - Attempting LDAP connection with ['ldap://ldap.example.com:389']
2021-07-24 05:36:29,450 - ldap_auth_provider - 471 - DEBUG - sentinel - Established LDAP connection in simple bind mode: ldap://ldap.example.com:389 - cleartext - user: ou=users,dc=cloudron - not lazy - unbound - closed - <no socket> - tls not started - not listening - SyncStrategy - internal decoder
```

Commenting out the entire section, works as expected i.e it doesn't connect to LDAP.

**Comment by gramakri** [comment]
OK, I think I found the issue.

* The `enabled` flag never had any effect. I was incorrect in my previous comment about it having any impact before 1.37
* The LDAP auth module previously used to error immediately with a bad server like ldap.example.com. In the recent one, it just hangs. I am not 100% sure but it seems to be related to async/await related changes in https://github.com/matrix-org/matrix-synapse-ldap3/commit/2e9d781f1eaec22a3fc84c7e5637683785083929. I will open an issue in that issue tracker.

**Comment by gramakri** [comment]
I opened https://github.com/matrix-org/matrix-synapse-ldap3/issues/124 for the invalid server address.
